src/generators/legacy-json/utils/buildSection.mjs

```diff
@@ -0,0 +1,385 @@
+import {
+  DEFAULT_EXPRESSION,
+  LEADING_HYPHEN,
+  NAME_EXPRESSION,
+  RETURN_EXPRESSION,
+  TYPE_EXPRESSION,
+} from '../constants.mjs';
+import { buildHierarchy } from './buildHierarchy.mjs';
+import parseSignature from './parseSignature.mjs';
+import { getRemarkRehype } from '../../../utils/remark.mjs';
+import { transformNodesToString } from '../../../utils/unist.mjs';
+
+const sectionTypePlurals = {
+  module: 'modules',
+  misc: 'miscs',
+  class: 'classes',
+  method: 'methods',
+  property: 'properties',
+  global: 'globals',
+  example: 'examples',
+  // Constructors should go under a class sections' signatures property
+  ctor: 'signatures',
+  classMethod: 'classMethods',
+  event: 'events',
+  var: 'vars',
+};
+
+/**
+ * @param {import('../types.d.ts').HierarchizedEntry} entry
+ * @returns {import('../types.d.ts').Meta | undefined}
+ */
+function createMeta(entry) {
+  const makeArrayIfNotAlready = val => (Array.isArray(val) ? val : [val]);
+  const { added_in, n_api_version, deprecated_in, removed_in, changes } = entry;
+  if (
+    !added_in &&
+    !n_api_version &&
+    !deprecated_in &&
+    !removed_in &&
+    changes.length < 1
+  ) {
+    return undefined;
+  }
+  return {
+    changes,
+    added: added_in ? makeArrayIfNotAlready(added_in) : undefined,
+    napiVersion: n_api_version
+      ? makeArrayIfNotAlready(n_api_version)
+      : undefined,
+    deprecated: deprecated_in
+      ? makeArrayIfNotAlready(deprecated_in)
+      : undefined,
+    removed: removed_in ? makeArrayIfNotAlready(removed_in) : undefined,
+  };
+}
+
+/**
+ * @param {import('../types.d.ts').HierarchizedEntry} entry Section's AST entry
+ * @param {HeadingMetadataParent} head Head node of the entry
+ * @returns {import('../types.d.ts').Section}
+ */
+function createSection(entry, head) {
+  const text = transformNodesToString(head.children);
+  return {
+    textRaw: text,
+    name: head.data.name,
+    type: head.data.type,
+    meta: createMeta(entry),
+    introduced_in: entry.introduced_in,
+  };
+}
+
+/**
+ * Find name, type, default, desc properties
+ * @param {import('mdast').ListItem} child
+ * @param {import('../types.d.ts').HierarchizedEntry} entry
+ * @returns {import('../types.d.ts').List}
+ */
+function parseListItem(child, entry) {
+  /**
+   * @type {import('../types.d.ts').List}
+   */
+  const current = {};
+
+  current.textRaw = child.children
+    .filter(node => node.type !== 'list')
+    .map(node =>
+      entry.rawContent.slice(
+        node.position.start.offset,
+        node.position.end.offset
+      )
+    )
+    .join('')
+    .replace(/\s+/g, ' ')
+    .replaceAll(/<!--.*?-->/gs, '');
+
+  if (!current.textRaw) {
+    throw new Error(`Empty list item: ${JSON.stringify(child)}`);
+  }
+
+  let text = current.textRaw;
+
+  // Extract name
+  if (RETURN_EXPRESSION.test(text)) {
```

Comment:
I'm usually against extremely long-function bodies, and I see a lot of repetition here. I bet this can become something generic and have a lot of stuff reused :)